src/monitors/ModuleMonitor.v3

```diff
@@ -0,0 +1,273 @@
+// Copyright 2023 Wizard Authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+class ModuleMonitor(module: Module) {
+	// Attach a callback that is called when func (non-import) is called
+	def beforeFuncCall(func: FuncDecl, f: DynamicLoc -> Resumption) {
+		if (func.imp != null) return;
+		beforeFuncCallAndReturn(func, f, null);
+	}
+	// Attach a callback that is called when func (non-import) returns
+	def beforeFuncReturn(func: FuncDecl, f: DynamicLoc -> Resumption) {
+		if (func.imp != null) return;
+		beforeFuncCallAndReturn(func, null, f);
+	}
+	// Attach callbacks that are called when a func is called or returns
+	def beforeFuncCallAndReturn(func: FuncDecl, callFn: DynamicLoc -> Resumption, retFn: DynamicLoc -> Resumption) {
+		var it = BytecodeIterator.new();
+		var bi = it.reset(func);
+
+		var startsWithLoop = false;
+		var frameAccessors: ListStack<FrameAccessor>;
+		if (bi.current() == Opcode.LOOP) {
+			startsWithLoop = true;
+			frameAccessors = ListStack<FrameAccessor>.new();
+			module.insertProbeAt(func.func_index, bi.pc, FuncWithLoopEnterProbe.new(frameAccessors, callFn));
+		} else if (callFn != null) {
+			module.insertProbeAt(func.func_index, bi.pc, CallbackProbe.new(callFn));
+		}
+
+		for (bi = it.reset(func); bi.more(); bi.next()) {
+			var op = bi.current();
+			match (op) {
+				RETURN => ;
+				END => if (bi.pc + 1 < bi.func.cur_bytecode.length) continue;
+				_ => continue;
+			}
+
+			if (startsWithLoop) {
+				module.insertProbeAt(func.func_index, bi.pc, FuncWithLoopExitProbe.new(frameAccessors, retFn));
+			} else if (retFn != null) {
+				module.insertProbeAt(func.func_index, bi.pc, CallbackProbe.new(retFn));
+			}
+		}
+	}
+	// Attach a callback to any non-import function call
+	def beforeCall(f: DynamicLoc -> Resumption) {
+		forEachFunc(MonitorUtil.isNotImport, beforeFuncCall(_, f));
+	}
+	// Attach a callback to any non-import function return
+	def beforeReturn(f: DynamicLoc -> Resumption) {
+		forEachFunc(MonitorUtil.isNotImport, beforeFuncReturn(_, f));
+	}
+	// Attach a callback to any non-import function call or return
+	def beforeCallAndReturn(callFn: DynamicLoc -> Resumption, retFn: DynamicLoc -> Resumption) {
+		forEachFunc(MonitorUtil.isNotImport, beforeFuncCallAndReturn(_, callFn, retFn));
+	}
+
+	// Attach a callback to an instruction
+	def beforeInstr(func: FuncDecl, pc: int, f: DynamicLoc -> Resumption) {
```

Comment:
Similarly, I think we could name this "beforeExecInstr" for symmetry and clarity.

Reply:
I ended up renaming them to `beforeFuncExec` and `beforeInstrExec` for symmetry with `beforeFuncReturn`, let me know if you think `beforeExecFunc` is better